src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Ricci.cpp

```diff
@@ -0,0 +1,167 @@
+      local_buffer(get_size(get<0, 0>(inverse_spatial_metric)));
+
+  // New variable to avoid recomputing in nested loops
+  auto& spatial_phi_Ijj =
+      get<::Tags::TempIjj<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t k = 0; k < VolumeDim; ++k) {
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = i; j < VolumeDim; ++j) {  // Symmetry
+        spatial_phi_Ijj.get(k, i, j) =
+            0.5 * inverse_spatial_metric.get(k, 0) * phi.get(0, 1 + i, 1 + j);
+        for (size_t l = 1; l < VolumeDim; ++l) {
+          spatial_phi_Ijj.get(k, i, j) +=
+              0.5 * inverse_spatial_metric.get(k, l) * phi.get(l, 1 + i, 1 + j);
+        }
+      }
+    }
+  }
+
+  // New variable to avoid recomputing in nested loops
+  auto& spatial_phi_ijK =
+      get<::Tags::TempijK<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t k = 0; k < VolumeDim; ++k) {
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = 0; j < VolumeDim; ++j) {
+        spatial_phi_ijK.get(i, j, k) =
+            0.5 * inverse_spatial_metric.get(k, 0) * phi.get(i, 1 + j, 1);
+        for (size_t l = 1; l < VolumeDim; ++l) {
+          spatial_phi_ijK.get(i, j, k) +=
+              0.5 * inverse_spatial_metric.get(k, l) * phi.get(i, 1 + j, 1 + l);
+        }
+      }
+    }
+  }
+
+  // New variable to avoid recomputing in nested loops
+  auto& spatial_phi_trace_second_third_indices =
+      get<::Tags::Tempi<0, VolumeDim, Frame, DataType>>(local_buffer);
+  auto& spatial_phi_trace_first_second_indices =
+      get<::Tags::Tempi<1, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t i = 0; i < VolumeDim; ++i) {
+    spatial_phi_trace_second_third_indices.get(i) =
+        spatial_phi_ijK.get(i, 0, 0);
+    spatial_phi_trace_first_second_indices.get(i) =
+        spatial_phi_Ijj.get(0, 0, i);
+    for (size_t j = 1; j < VolumeDim; ++j) {
+      spatial_phi_trace_second_third_indices.get(i) +=
+          spatial_phi_ijK.get(i, j, j);
+      spatial_phi_trace_first_second_indices.get(i) +=
+          spatial_phi_Ijj.get(j, j, i);
+    }
+  }
+
+  auto& spatial_phi_trace_second_third_indices_I =
+      get<::Tags::TempI<0, VolumeDim, Frame, DataType>>(local_buffer);
+  auto& spatial_phi_trace_first_second_indices_I =
+      get<::Tags::TempI<1, VolumeDim, Frame, DataType>>(local_buffer);
+
+  raise_or_lower_index(make_not_null(&spatial_phi_trace_second_third_indices_I),
+                       spatial_phi_trace_second_third_indices,
+                       inverse_spatial_metric);
+  raise_or_lower_index(make_not_null(&spatial_phi_trace_first_second_indices_I),
+                       spatial_phi_trace_first_second_indices,
+                       inverse_spatial_metric);
+
+  for (size_t i = 0; i < VolumeDim; ++i) {
+    for (size_t j = i; j < VolumeDim; ++j) {  // Symmetry
+      ricci->get(i, j) = 0.;
+      for (size_t k = 0; k < VolumeDim; ++k) {
+        ricci->get(i, j) +=
+            0.5 *
```

Comment:
There is a minus sign applied to both terms here as compared to the documentation (so the math does match, it's just harder to compare) -- it'll be easier to trace through this in the future if they match.

Reply:
changed